deployment: adds content (#28)
Ada-Developers-Academy/core-unit-2: deployment/deployment.md

# Deployment

## Learning Goals

- Define software deployment

## Introduction

Our applications and web APIs are developed locally and available locally—they are only accessible while our local machines are running a server, and only through `localhost`.

If our personal machines are turned off, then our server isn't running, and our API isn't accessible.

> How do we make our web apps run online reliably?

Reviewer: I think this sounds a bit more exciting
```suggestion
How do we make our web apps available to the rest of the world?
```

## Vocabulary and Synonyms

| Vocab      | Definition                                          | Synonyms            | How to Use in a Sentence                                                                                                                                         |
| ---------- | --------------------------------------------------- | ------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Deployment | The process of making software available to be used | Software deployment | "After I deploy this software, I can send my deployed app to my friends so they can use it!" "I can deploy software updates so my deployed app grows over time." |

## What is Deployment?

Software deployment is the process of making some software or application available for use, often to an end-user.

> If we _deploy_ our apps, we will set up machines and web servers dedicated to running our web app. Then, users can access our apps without relying on our own personal machines!

Reviewer: "when", not "if"
"relying" sounds odd to me here. This part is less important.
```suggestion
When we _deploy_ our apps, we will set up machines and web servers dedicated to running our web app. Then, users can access our apps regardless of what our own personal machines are doing!
```

We deploy apps using skills and tools in the realm of:

- DevOps (Software Development and IT Operations)
- Systems administration
- Database administration

These areas introduce the following kinds of questions:

- What kind of machine is running our web servers?
- What is installed on these machines? How do they manage dependencies?
- How do we push new code to these servers?
- How do we keep these servers available?
- How do we maintain these servers, when errors arise?

> In this curriculum, we will choose specific deployment tools and services that answer these questions for us, so that we may focus on web development skills.

Reviewer: we are expressing an ability, not a need for permission
I'd really like to say "focus on developing our web development skills" but that's too redundant-sounding to me (even though it has a slightly different connotation to me than "improving")
```suggestion
In this curriculum, we will choose specific deployment tools and services that answer these questions for us, so that we can focus on improving our web development skills.
```

### Example Platforms

> The following are all examlpes of _Platforms as a Service_. The responsibility of these services is to create, manage, and maintain environments that run our deployed apps.

Reviewer: ```suggestion
The following are all examples of _Platforms as a Service_. The responsibility of these services is to create, manage, and maintain environments that run our deployed apps.
```

- AWS Elastic Beanstalk
- Heroku
- Windows Azure
- Google App Engine

### !callout-info

## Deploy to the Production Environment

> Beyond the "development environment" and the "test environment," we can introduce a third "production environment." This production environment refers to the environment that runs our deployed, released application.

Reviewer: can we echo back to the dancing analogy?
```suggestion
The _production_ environment is like Aleesha's main stage! It refers to the environment that runs our deployed, released application. This is where all our hard work gets made available to actual users!
```

> The _production_ environment is like Aleesha's main stage! It refers to the environment that runs our deployed, released application. This is where all our hard work gets made available to actual users!

Reviewer 2: This looks like a person's name, but I'm confused by the reference to a person here when I don't see their name anywhere else.

Reviewer: Aleesha's story is introduced in the flask testing lesson to frame the discussion about environments. She's a dancer who dances at home, at a practice hall, and in performances on the main stage. 🙂

Reviewer 2: Oh, got it! This seems to break the guidance I've heard about not requiring an ordering to Learn topics, but if that guidance has been dropped, totally cool with me

Reviewer: Hmm. That does make sense... On the other hand, it seems a shame to miss out on the opportunity to tie this directly into the metaphoric framework that was established earlier for reinforcement.

Lessons can't exist entirely in a vacuum since 1. there are concepts that build on each other (deploying to the production environment depends on the concept a production environment) and 2. it can sound super tone deaf to talk about a topic that a reader _has_ encountered already as though they haven't seen it before.

This is only a side comment, so I won't fight dropping this if desired, but at the same time, we have to allow for _some_ assumption of ordering.

Reviewer 2: Totally hear the tradeoff. I think that will always be the downside of not implying ordering for the topics. 

If the concern with implying ordering is that we want to be able to reorder them without accidentally missing stuff like this, I wonder if we could solve that by just keeping an instructor document somewhere (maybe one for each learn repo, in the root) that notes lessons that contain things that depend on previous lessons.

Author: Personally I like this callout especially because the Aleesha metaphor is only brought up in the previous Testing topic!

I think that an instructor or someone with a voice could bring it back to Aleesha; I'm gonna go with a change to smooth this out (sorry metaphorical dance crew!) and then move forward with this PR

```suggestion
The _production_ environment refers to the environment that runs our deployed, released application. This is where all our hard work gets made available to actual users!
```

Reviewer: Aleesha, bye! ✋

### !end-callout

## When Do We Deploy?

We deploy and re-deploy apps whenever we want to release a new version of our application.

On large software products, we may deploy according to a release schedule.

### !callout-info

## Continuous Deployment

_Continuous deployment_ is a software engineering practice that values and encourages deploying frequently, favoring frequent, small updates over large, cumbersome ones. This practice gives us the phrase, "Deploy early and often!"

### !end-callout

## Our New Dev Workflow

> Assuming we want to deploy early and often, our new development workflow may look like this:
> 
> 1. Begin a new feature
> 1. Do work in our local environment, using best practices for git and testing
> 1. Finish the new feature
> 1. Deploy to the production environment

Reviewer 2: I'm thinking naming the implied context of an existing application would be useful here. Once I added that context, the word "new" felt a little awkward to me, but maybe that's just me.
```suggestion
Let's say we have an existing web application that is already deployed and already being used by people all over the world. Assuming we want to deploy early and often, our development workflow may look like this:

1. Begin a new feature
1. Do work in our local environment, using best practices for git and testing
1. Finish the new feature
1. Deploy the updated code with this new feature to the production environment
```

Reviewer: I like this change too. 👍
